Python/Product/Ipc.Json/Connection.cs

```diff
@@ -556,30 +556,67 @@ CancellationToken cancel
-                    await _writer.WriteAsync(contentBytes, 0, contentBytes.Length).ConfigureAwait(false);
-                    await _writer.FlushAsync(cancel).ConfigureAwait(false);
-                } finally {
-                    _writeLock.Release();
+                while (true) {
+                    try {
+                        await _writeLock.WaitAsync().ConfigureAwait(false);
+                    } catch (ArgumentNullException) {
+                        throw new ObjectDisposedException(nameof(_writeLock));
+                    } catch (ObjectDisposedException) {
+                        throw new ObjectDisposedException(nameof(_writeLock));
+                    }
+                    try {
+                        sentSeq = Interlocked.CompareExchange(ref _sentSeq, seq, seq - 1);
+                        if (sentSeq == seq - 1) {
+                            cancel.ThrowIfCancellationRequested();
+
+                            // The content part is encoded using the charset provided in the Content-Type field.
+                            // It defaults to utf-8, which is the only encoding supported right now.
+                            var contentBytes = TextEncoding.GetBytes(str);
+
+                            // The header part is encoded using the 'ascii' encoding.
+                            // This includes the '\r\n' separating the header and content part.
+                            var header = "Content-Length: " + contentBytes.Length + "\r\n\r\n";
+                            var headerBytes = Encoding.ASCII.GetBytes(header);
+
+                            await _writer.WriteAsync(headerBytes, 0, headerBytes.Length).ConfigureAwait(false);
+                            await _writer.WriteAsync(contentBytes, 0, contentBytes.Length).ConfigureAwait(false);
+                            await _writer.FlushAsync().ConfigureAwait(false);
+                            return;
+                        } else if (sentSeq < seq - 1) {
+                            // Not our turn to send yet
+                        } else if (sentSeq == seq) {
+                            // Uh... what happened here?
+                            Debug.Fail($"Message {seq} has already been sent");
+                            LogToDisk($"Message {seq} has already been sent");
+                            break;
+                        } else if (sentSeq > seq) {
+                            // We were missed
+                            Debug.Fail("Message was not sent");
+                            LogToDisk($"Message {seq} was skipped because message {sentSeq} has already been sent");
+                            break;
+                        }
+                    } finally {
+                        _writeLock.Release();
+                    }
+
+                    --retries;
+                    if (retries > 100) {
+                        // Fast looping
+                        await Task.Yield();
```

Comment:
I don't like this approach. We have no explicit message pool, so with multiple events sent concurrently, we can end up sending these messages forever. Also, since there is no guarantee that this code is executed in background thread, this loop can run on UI thread.

On the other hand, `_sentSeq` doesn't change in normal situation outside of the lock, so it is either correct in the first loop iteration or it will never be correct for that message again unless there is another message with incorrect `seq`.

Why can't we just assing `ProtocolMessage.seq` in the `SendMessage` after `_writeLock` is acquired?

Reply:
Because this starts running on the thread that's sending the message, we can have multiple interleaving contexts in this loop at once. An alternative would be to always create a new thread for each connection and serialize messages through that (i.e. use a queue that the worker thread pulls from), and that may still turn out to be necessary. But this approach certainly works for the core case (multiple threads hit the lock, then get released in random order).

The messages themselves need to be in order when posted from the same thread, and by assigning seq synchronously we can ensure that. If we assign it later, then the messages are still out of order, but we've lost the only useful piece of information that can prove it.

Reply:
Unless we create tasks in a single-threaded synchronization context, we have no guarantee that two code pieces, separated by `await` operator, will be executed in the same thread. Increasing `Connection._seq` twice inside asynchronous code flow doesn't guarantee that numbers will be sequential if there is another asynchronous code flow.

If we want to ensure that several events are sent without interruption, we should either provide access to some shared lock that can be acquired by event-raising, or collect events in a batch before sending. Current approach only ensures that events are sent in the same order they are created, but it doesn't guarantee that they are created in the right order.

This code may create problems when queue of events is too long. For instance, if there are 3 events that come to the lock in order `1,3,2`, third event will be waiting on the lock while the first is processed. If during this waiting 10 more events arrive, they all will be put on the tail of the lock and after first event is processed third one will end up in the tail of this queue. `SemaphoreSlim`, at least in its current implementation, execute continuations in the same order they are scheduled, but `Task.Yield` doesn't preserve any order, so on the second loop of every element in the queue they all will be shuffled, even if originally they were in the right order. As a result, we get a bubble sort with an expense of creating 2 tasks for every iteration of every message that isn't in the right place.

It may completely fail if event with expected `seq` never arrives. To avoid this in case of cancellation, `CancellationToken` isn't passed into `WaitAsync` anymore, which means that task can't be cancelled unless it is actually processed. But if, for instance, `JsonConvert.SerializeObject` can fail with exception, `_sentSeq` will become unsynced with `_seq`.

Reply:
"In the same thread" doesn't matter here - we have a lock - and numbers only have to be increasing, not sequential (that is, if thread A sends message M1 and M2 and B sends M3, as long as M2 comes after M1 it doesn't matter if M3 is in the middle).

`Task.Yield` is only going to have one continuation, so I don't see how things get shuffled. The semaphore is where they all collect, and when it becomes available then one will be released (in order of arrival, as you said). If a continuation is executed, doesn't match the expected seq, and _does not_ yield, it will release and immediately reacquire the lock, repeating until it exhausts its retries (validated by testing).

Good point about not handling the exceptional case. I'll handle that.

Reply:
> if thread A sends message M1 and M2 and B sends M3, as long as M2 comes after M1 it doesn't matter if M3 is in the middle

Ok, then I don't understand how is it possible that M2 comes before M1. There are only 4 methods that call `SendMessage`:  `SendRequestAsync`, `SendEventAsync`, `SendResponseAsync` and `WriteError`. For all of them, code flow piece between incrementing the `_seq` and calling `ConfigureAwait(false)` on a task returned from the `_writeLock` is 100% synchronous. And `SemaphoreSlim` is FIFO, it will preserve the order.

So as far as I understand it, current approach doesn't fix that M2 always comes after M1, cause it is happening already, it enforces that M3 comes after M2, while we actually doesn't care about it.

> so I don't see how things get shuffled. 

I've created a standalone simulation test. It sends 10 ints in order `0,2,1,3,4,5,6,7,8,9` to the following loop:

```cs
while (true) {
    var lockTask = _lock.WaitAsync();
    AddTaskToSequence(lockTask.Id, index);
    await lockTask;
    RemoveTaskFromSequence(lockTask.Id);
    try {
        if (Interlocked.CompareExchange(ref _index, index, index - 1) == index - 1) {
            TraceSequence(index);
            await Task.Delay(1000);
            return;
        }
    } finally {
        _lock.Release();
    }
    await Task.Yield();
}
```

I've also added another method that schedules short work items into thread pool to simulate real application environment.

Here is what I got:
- `0` is processing, queue is `2,1,3,4,5,6,7,8,9`
- `2` is skipped and put in the tail of the queue
- `1` is processing, queue is `3,4,5,6,7,8,9,2`
- All the queue is skipped until `2` is reached. At this point, every item in the queue (except `2`) reaches `Task.Yield()` and their continuations are executed in random order
- `2` is processing, queue is `3,5,4,8,6,7,9`
- `3` is processing, queue is `5,4,8,6,7,9`
- `5` is skipped and put in the tail of the queue
- `4` is processing, queue is `8,6,7,9,5`
- All the queue is skipped until `5` is reached.
- `5` is processing, queue is `8,6,9,7`

And so on.

Reply:
That doesn't look like shuffling to me - it's just looping through the tasks until it finds the next one (insertion sort, perhaps?).

I'm not convinced that `SemaphoreSlim` is reliably FIFO, as it relies on `lock()` statements upon entry to its methods, but if it is then we should just get optimal performance in most cases. You haven't suggested an alternative approach yet (other than assuming that everything will be okay, which we've already seen doesn't work :) ), but if you have an idea I'd love to hear it.

Reply:
> I'm not convinced that `SemaphoreSlim` is reliably FIFO, as it relies on `lock()` statements upon entry to its methods

Locks don't matter. If `_seq` increment for M2 can happen only after `SendRequestAsync`, `SendEventAsync`, `SendResponseAsync` or `WriteError` returns its task, then it is guarantee to happen after first synchronous part is completed. That synchronous part can wait on the lock, but it is a synchronous wait, so method won't return in this case. `SemaphoreSlim` is reliably FIFO unless someone changes that implementation. This is a simple experiment:

```cs
        private static SemaphoreSlim _ss = new SemaphoreSlim(1);
        private static int _value = -1;

        static void Main(string[] args) {
            var tasks = new Task[100];
            for (int i = 0; i < 100; i++) {
                tasks[i] = AddToSemaphore(i);
            }
            Console.WriteLine("Done scheduling");
            Task.WaitAll(tasks);
            Console.ReadKey();
        }

        private static async Task AddToSemaphore(int i) {
            await _ss.WaitAsync().ConfigureAwait(false);
            await Task.Delay(100);
            Console.Write($"{i},");
            if (Interlocked.Exchange(ref _value, i) != i - 1) {
                Console.WriteLine();
                Console.WriteLine($"Wrong order!");
            }
            _ss.Release();
        }
```

Note that if you add `await Task.Yield();` at the beginning of the `AddToSemaphore`, you'll get several errors.

> You haven't suggested an alternative approach yet (other than assuming that everything will be okay, which we've already seen doesn't work :) )

I suggest that we find the real cause of the issue. If it is guaranteed that message M2 won't be created until synchronous part of the method that created previous message M1 is completed, messages will be sent in order they are created. So the issue is somewhere else:
- Maybe some of them weren't sent because of the cancellation token, and receiving side doesn't expect that.
- Maybe they aren't created in the expected order.
- Maybe they are actually created in different threads.
- Maybe we have more than one instance of connection.

If you share what are the repro steps and symptoms of the issue, I can take a look.

Reply:
I started writing out the exact sequence and proved to myself that you're right :) I'll revert this part of the change and reanalyze the likely cause.

The repro steps are basically to edit Python code for a period of time, making edits, and then the buffer gets out of sync. When that happens, there are typically two fileUpdate messages (in the connection logs - see the wiki for enabling these): one resets to version 0 and then the next applies a change to the version prior to the reset (i.e. the reset preceded the update). I'll try and track it down.